elastic/integrations
# Add Firehose integration for logs

## What does this PR do?

This PR is to add firehose integration into AWS package. This integration supports reroute processors to provide routing configurations for logs. For example, if the field `aws.cloudwatch.log_stream` value contains `CloudTrail`, then route this log entry from the default `logs-aws.firehose_logs-default` to `logs-aws.cloudtrail-default` data stream. 

The original plan is to add firehose as an input package outside the existing AWS package. But package spec does not support predefined pipelines in input packages nor routing rules. Please see https://github.com/elastic/package-spec/issues/566 for more details. 

### Problems with adding `firehose` inside AWS integration
1. When a user adds `firehose` integration, the initial page they see on Kibana is the documentation for the AWS package, not specific to Firehose. (This can be fixed if we create a separate integration package for Firehose).
<img width="1230" alt="Screenshot 2023-07-17 at 3 22 12 PM" src="https://github.com/elastic/integrations/assets/14081635/4a84f6b1-adfb-4a41-816b-1d8a65f61058">

2. When click `Add Firehose`, the page for adding configurations pops up. But for firehose integration, we do not need users to do any configuration setup so this will confuse the users and also make it less friendly. (This can be fixed if we create a separate integration package for Firehose).
<img width="787" alt="Screenshot 2023-07-17 at 3 40 59 PM" src="https://github.com/elastic/integrations/assets/14081635/fed77e85-7ae2-4533-ac5e-7375d6a983d4">

3. When in the future if we add support for metrics, we need a separate data stream called `firehose_metrics`. I can't define a data stream called `firehose` and has both `type: metrics` and `type: logs` inside.

### Benefits of having `firehose` inside AWS integration:
1. Users don't need one extra step to install `aws` integration separately. Because once `firehose` is added, all the assets(ingest pipelines especially) are installed automatically. Note: maybe we should make an ER to Fleet regarding installing additional assets/packages. 

## Checklist


## How to test it
1. Install firehose integration
2. You can check in Kibana for ingest pipelines to see if the reroute processors are there in `logs-aws.firehose_logs-<version>`
3. In dev tool, send ES a test document and make sure it's stored into the default data stream `logs-aws.firehose_logs-default`. For example:
```
  POST /logs-aws.firehose_logs-default/_doc
  {
    "cloud.region": "us-east-2",
    "aws.firehose.arn": "arn:aws:firehose:us-east-2:123456789:deliverystream/firehose-cloudtrail-logs-to-elastic",
    "data_stream.namespace": "default",
    "aws.firehose.subscription_filters": "[cloudtrail-to-firehose]",
    "message": "{\"eventVersion\":\"1.08\",\"userIdentity\":{\"type\":\"AWSService\",\"invokedBy\":\"cloudtrail.amazonaws.com\"},\"eventTime\":\"2023-07-17T21:02:26Z\",\"eventSource\":\"sts.amazonaws.com\",\"eventName\":\"AssumeRole\",\"awsRegion\":\"sa-east-1\",\"sourceIPAddress\":\"cloudtrail.amazonaws.com\",\"userAgent\":\"cloudtrail.amazonaws.com\",\"requestParameters\":{\"roleArn\":\"arn:aws:iam::123456789:role/service-role/aws-cloudtrail-logs-123456789-b888baff_Role\",\"roleSessionName\":\"CLOUDWATCH_LOGS_DELIVERY_SESSION\"},\"responseElements\":{\"credentials\":{\"accessKeyId\":\"TEST\",\"sessionToken\":\"TEST\",\"expiration\":\"Jul 12, 2023, 10:02:26 PM\"},\"assumedRoleUser\":{\"assumedRoleId\":\"AROAZEDJODE3NLJAH2FZC:CLOUDWATCH_LOGS_DELIVERY_SESSION\",\"arn\":\"arn:aws:sts::123456789:assumed-role/aws-cloudtrail-logs-123456789-b888baff_Role/CLOUDWATCH_LOGS_DELIVERY_SESSION\"}},\"requestID\":\"041c9e5f-a031-47d2-a4a0-011bc8d5352c\",\"eventID\":\"3096b662-7aa9-43e6-8bee-541a45686745\",\"readOnly\":true,\"resources\":[{\"accountId\":\"123456789\",\"type\":\"AWS::IAM::Role\",\"ARN\":\"arn:aws:iam::123456789:role/service-role/aws-cloudtrail-logs-123456789-b888baff_Role\"}],\"eventType\":\"AwsApiCall\",\"managementEvent\":true,\"recipientAccountId\":\"123456789\",\"sharedEventID\":\"a1c94275-884f-4c1f-b8dc-2e1bf4c94d29\",\"eventCategory\":\"Management\"}",
    "aws.kinesis.type": "deliverystream",
    "data_stream.type": "logs",
    "aws.firehose.request_id": "971ae05f-a128-4a7f-b623-30f9bc513e55",
    "aws.cloudwatch.log_stream": "123456789_CloudTrail_us-east-2_3",
    "cloud.provider": "aws",
    "@timestamp": "2023-07-17T21:04:35Z",
    "cloud.account.id": "123456789",
    "data_stream.dataset": "aws.firehose_logs",
    "aws.kinesis.name": "firehose-cloudtrail-logs-to-elastic",
    "event.id": "37670326805251200781477669690942747782212394134076063744",
    "aws.cloudwatch.log_group": "aws-cloudtrail-logs-123456789-1c167310"
  }
```
5. Check in Discover to see what's the document looks like. It should be under `logs-aws.cloudtrail-default` instead and with message field all parsed by the `cloudtrail` ingest pipelines.

## Screenshots
<img width="1225" alt="Screenshot 2023-07-17 at 3 21 59 PM" src="https://github.com/elastic/integrations/assets/14081635/acfb1f7e-a631-4eb7-935b-e18b6f4ef6a8">
